### DISCUSSION TITLE
How to set hostname properly?

### DISCUSSION THREAD

**Discussion by cziter15** [esp8266/Arduino#8816] [Q&A] [answered]
Hello,

How to trigger hostname setting properly using WiFi class / station mode?
Should setHostname be called before WIFI_STA mode set (or begin)? Or after?

I'm playing with this functionality and getting mixed results. Why i have to call wifi_station_set_hostname to get it working?
```
WiFi.mode(WIFI_STA);
WiFi.setHostname(hostname);
wifi_station_set_hostname(hostname);
```

Without calling wifi_station_set_hostname or changing the order, it fallbacks to ESP-XXXXX hostname. Looks like something broken recently.

**Comment by mcspr** [comment]
Have you checked whether `setHostname(...)` returns `true` or tried enabling debugging?
What hostname string are you trying to set?

I see that our hostname func might be picky about format. But, it does not depend on interface being up
https://github.com/esp8266/Arduino/blob/39b41379ee5bc08b907179d3bcd71a1591dbc365/cores/esp8266/LwipIntf.cpp#L119-L198

Using SDK func just goes around that filtering, but you have to enable WiFi before it works (unlike our method)

**Comment by cziter15** [comment]
Need to debug it, but to do it I need  to get a board on my desk. Now the device is mounted.

Anyway, yes I've read that code. It worked earlier, but after updating arduino and switching to sdk 3.0.5 the behavior has changed.

The name is Raesp in this case, it should work.

**Comment by mhightower83** [comment]
In the past, I have also had this issue. It initially wasn't clear to me what I did that fixed it. I now think these are some additional key components to make it work:
* I think it is key to start from the `WL_IDLE_STATUS` state. Make sure the ESP is not using remembered SSID and password.
* `WiFi.hostname(aHostname)` must precede `WiFi.begin()`
* The Router needs to support registering DHCP host names from clients into the active local DNS resolver.

Also for me, it works with SDK v3.0.5.

**Comment by cziter15** [comment]
Well this is very strange how it behaves.

My current implementation:
```
  // somewhere at initialization
  WiFi.mode(WIFI_STA);
  hostNameChanged = WiFi.setHostname("Raesp");
  setupMacAddress();
  WiFi.setAutoConnect(false);
  WiFi.setAutoReconnect(false);

  // post initialization
  WiFi.begin();
```

Then asking for hostname (later):
```
respond(
	PGM_("HNC: ") + std::to_string(hostNameChanged ) + 
	PGM_(", HN: ") + WiFi.getHostname()
);
```
and it responds
`HNC: 1, HN: ESP-7800CB`.

Well it says that the setHostname returned true, but when asking for hostname it remains unchanged/default.
Anyway, I'm unable to erase SSID/PWD as I'm using this feature (this is how WiFiManager handles provisioning persistence).

It looks like some other function overrides that hostname... but what function? Maybe esp sdk?

**Reply by mcspr** [reply]
Do you see this behaviour without WiFiManager? And we'll have to know what WiFimanager uses for persistence, perhaps that also matters.
*(note that Github discussions, unlike issues, are threaded :)*

Similar to https://github.com/esp8266/Arduino/discussions/8816#discussioncomment-4723724, I don't observe hostname changing to default at any point.
loop() with Serial.println(WiFI.hostname()) keeps reporting the same thing for a basic sketch with mode() and WiFi.begin(SSID, AUTH)

**Reply by cziter15** [reply]
I'll need to prepare somehow and dig into this issue deeper.

For now, using this piece of code:
```
WiFi.mode(WIFI_STA);
WiFi.setHostname(hostname);
wifi_station_set_hostname(hostname);
```
it seems that  WiFi.getHostname() returns proper hostname (looks like wifi_station_set_hostname is critical), but this is strange....

When it comes to persistence... How it's affected? I can't find anything related to hostname in station_config.
Moreover, it worked properly a month ago or so (before switching to v3.0.5, but I need to test it).

**Reply by mcspr** [reply]
Depends on where we lose the hostname. Note that using WiFi methods should be sufficient for testing; mixing in SDK funcs just for the sake of using them may not produce the consistent and easy to understand result for us (since, like mentioned above, we are not able to reproduce just yet)

In case of WiFimanager, I don't see anything extra done to hostname(s) there. But... it also has its own setHostname() override being used at setup, which in turn would override whatever being set with WiFi.hostname(...) by you. A more simple explanation than hostname being saved & restored somewhere, but idk how the rest of you code works.

Also note that 3.0.5 is pretty new for us, so maybe we missed some condition that is resetting it.

**Reply by cziter15** [reply]
Yeah, looks like without more info we arr unable to figure out what goes wrong. The only thing that for now I know, is that something changed ;)

I have also ArduinoOTA component and it looks like mdns is working only for some time (device is visible as OTA port only for a minute or less just agter reboot).

**Reply by cziter15** [reply]
It looks like that calling wifi_set_macaddr overrides hostname. I've changed it and now I'm setting MAC before applying hostname it seems to work.

When it comes to OTA & hostname, it looks like MDNS is not working properly for some reason.

**Reply by mcspr** [reply]
> It looks like that calling wifi_set_macaddr overrides hostname. I've changed it and now I'm setting MAC before applying hostname it seems to work.

wow, ok :)

```console
> xtensa-lx106-elf-gdb --batch --ex "disassemble wifi_set_macaddr" .pio/build/d1_mini/firmware.elf
...
  0x4023ec43 <+283>:   call0   0x4023d96c <wifi_station_set_default_hostname>
  0x4023ec46 <+286>:   j       0x4023ec33 <wifi_set_macaddr+267>
```
jumps here right at the end of some kind of loop comparing something in netif. did not happen in older SDKs, referencing libmain.a:user_interface.o

**Reply by cziter15** [reply]
About... mDNS, I'll try to look at this, looks like it stops responding after a while.

**Reply by cziter15** [reply]
When it comes to MDNS, it looks like after latest changes it must be started AFTER connecting to the WiFi. Otherwise it stops working after a while (probably after TTL expiration).

**Reply by d-a-v** [reply]
I moved mDNS init code from the [clock example](https://github.com/esp8266/Arduino/blob/bed2fa339d1d6cfd281c06594431f19096d1ca10/libraries/ESP8266mDNS/examples/LEAmDNS/mDNS_Clock/mDNS_Clock.ino#L231-L240) to just after `Serial.begin()` and I still can see the service responsive and running 10 minutes after.

Can you provide a way to reproduce your issue ?

**Reply by cziter15** [reply]
I'm not calling MDNS.announce() like clock example does (line 264). Is this a problem? ArduinoOTA doesn't call this function too.

Well, the issue originally appeared when using ArduinoOTA with built-in MDNS support. I was able to workaround this issue, by adding onConnected/onDisconnected callbacks in my framework and then triggering all mdns logic manually.

I can try to write some small sketch to reproduce the issue, but currently everything is glued into my iot library built on top of Arduino.

The code below is a code of ota updater component, and it works properly after listed fixes.
```
/*
 *	Copyright (c) 2021-2023, Krzysztof Strehlau
 *
 *	This file is part of the ksIotFramework library.
 *	All licensing information can be found inside LICENSE.md file.
 *
 *	https://github.com/cziter15/ksIotFrameworkLib/blob/master/LICENSE
 */

#if ESP32
	#include <WiFi.h>
	#include <ESPmDNS.h>
#elif ESP8266
	#include <ESP8266WiFi.h>
	#include <ESP8266mDNS.h>
#else 			
	#error Platform not implemented.
#endif

#include "../ksApplication.h"
#include "../ksConstants.h"
#include "ksWifiConnector.h"
#include "ksOtaUpdater.h"

namespace ksf::comps
{
	ksOtaUpdater::ksOtaUpdater()
		: ksOtaUpdater(PGM_("ota_ksiotframework"))
	{}

	ksOtaUpdater::ksOtaUpdater(const std::string& password)
	{
		ArduinoOTA.setPassword(password.c_str());

		ArduinoOTA.onStart([&]() {
			onUpdateStart->broadcast();
		});

		ArduinoOTA.onEnd([&]() {
			ksf::saveOtaBootIndicator();
			onUpdateEnd->broadcast();
		});
	}

	bool ksOtaUpdater::init(ksApplication* owner)
	{
		/* Bind to WiFi events. */
		if (auto wifiConnSp{owner->findComponent<ksWifiConnector>().lock()})
		{
			wifiConnSp->onConnected->registerEvent(wifiConnEventHandleSp, std::bind(&ksOtaUpdater::onWifiConnected, this));
			wifiConnSp->onDisconnected->registerEvent(wifiDisconnEventHandleSp, std::bind(&ksOtaUpdater::onWifiDisconnected, this));		
		}

		/* Initialize Arduino OTA. */
		#if ESP8266
			ArduinoOTA.begin(false);
		#elif ESP32
			ArduinoOTA.setMdnsEnabled(false);
			ArduinoOTA.begin();
		#else 
			#error Platform not implemented.
		#endif

		return true;
	}

	void ksOtaUpdater::onWifiConnected()
	{
		MDNS.begin(WiFi.getHostname());
		isWifiAlive = true;
	}

	void ksOtaUpdater::onWifiDisconnected()
	{
		isWifiAlive = false;
		MDNS.end();
	}

	bool ksOtaUpdater::loop()
	{
		/* Handle OTA stuff. */
		ArduinoOTA.handle();

		/* Handle MDNS stuff. */
		#if ESP8266
			if (isWifiAlive)
				MDNS.update();
		#endif
		
		return true;
	}
}
```

**Reply by d-a-v** [reply]
`ArduinoOTA` calls [`MDNS.update()`](https://github.com/esp8266/Arduino/blob/cf240242926a2eeefde44b6c6218208918db321f/libraries/ArduinoOTA/ArduinoOTA.cpp#L373) (this call was added after many OTA users not calling it, and I wonder whether it is a mistake to have made this call hidden).
This should always be done when using MDNS.

Did your issue disappear after adding this call in the main loop ?

**Reply by cziter15** [reply]
I have that update call in my loop. I was thinking about announce function, that's called by example you've linked above. It appears to be called every second.

https://github.com/esp8266/Arduino/blob/bed2fa339d1d6cfd281c06594431f19096d1ca10/libraries/ESP8266mDNS/examples/LEAmDNS/mDNS_Clock/mDNS_Clock.ino#L264

**Reply by d-a-v** [reply]
OK with `update()` which is the right one to call.
`announce()` is used to announce a new service (new value for time in this example). 

> The code below is a code of ota updater component, and it works properly after listed fixes.

So this code of yours is working properly, but not another one of yours ?

Did you try the examples ?

Did you try to enable debug output with the non-working code ?

**Reply by cziter15** [reply]
It looks like that behavior is triggered by calling logic that blocks loop function.
I have something like that inside loop function. Don't know why it triggered mDNS issue now as it worked properly earlier....

```
void loop()
{
  while (1)
  {
     delay(1);
     whole_app_logic();
  }
}
```

I've been always thinking that delay() or yield() is enough to satisfy other tasks.

**Reply by d-a-v** [reply]
Is this code working with esp8266 arduino core v3.0.2 but not with v3.1.1 ?

**Reply by cziter15** [reply]
It was working earlier, I don't know precisely when, but it was few (Arduino) releases ago. I've been using default platformio SDK. That means default NONOS. On ESP32 same logic always worked properly, but yeah, it's RTOS there.

**Reply by d-a-v** [reply]
Can you try #8825 with your application and tell whether mDNS is better ?

**Reply by cziter15** [reply]
Will try later today. I've fixed my library so now it doesn't block loop anymore. Everything seems to be working properly now.

Is it allowed from Arduino-esp8266 perspective to "expropriate" the loop function and call own while loop with yield/delay inside or not? What about writing own main?

**Reply by d-a-v** [reply]
Then do not try #8825, it breaks the protocol which is otherwise working without this patch.
Thanks for your report !

**Reply by cziter15** [reply]
I've flashed my boards yesterday and currently using ArduinoOTA.begin with hostname (everything before actually connecting to WiFi) and mDNS is properly responding (>10h uptime).

@d-a-v, Are you able to answer the question about blocking inside loop? A lot of people do delays inside loop, which may break mDNS respoder as we see right now.

**Reply by d-a-v** [reply]
On esp8266, blocking inside a loop will not allow services to work properly.
Any service needing a regular call like `.update()`, `.loop()`, `.handleClients()` and also WiFi and mandatory IP services can be impacted, especially when this kind of loop is done inside an external library not aware of other existing services.

Using [recurrent scheduled functions](https://github.com/esp8266/Arduino/blob/a9675ca837609e4c3e9118f7ae74e176402403ca/cores/esp8266/Schedule.h#L71-L89) might help since core version 3.1 as they are now also called from `delay(n)` but always in "cont" mode (= not in system or OS or interrupt mode = anything except `delay()`/`yield()` is allowed).
They can be used [like Ethernet does](https://github.com/esp8266/Arduino/blob/a9675ca837609e4c3e9118f7ae74e176402403ca/cores/esp8266/LwipIntfDev.h#L276-L282).
We've not tried yet to extensively use them for our core services (except for Ethernet), but you can try with mDNS and report if it is working for you.

Note: this is not portable and very specific to esp8266 arduino core.

**Reply by cziter15** [reply]
Thank you for the answer 👍. I've clossed this discussion with tl;dr as everything asked seems to be answered.

**Comment by hmueller01** [comment]
I was struggling here too.
I use
```
  Serial.println(WiFi.getHostname());
  WiFi.mode(WIFI_STA);
  WiFi.setHostname("myESP");
  ...
  WiFi.begin();
  Serial.println(WiFi.getHostname());
  ArduinoOTA.begin();
  Serial.println(WiFi.getHostname());
```
I get
```
ESP-6A4C81
myESP
myESP
```
But in my router I see
`esp8266-6a4c81`
Why is ArduinoOTA not using the `WiFi.setHostname`?
```
  if (!_hostname.length()) {
    char tmp[15];
    sprintf(tmp, "esp8266-%06x", ESP.getChipId());
    _hostname = tmp;
  }
```
if not explicit set with `ArduinoOTA.setHostname("myESP");`?

**Reply by cziter15** [reply]
This is normal, esp32 implementation works same way. If you want ArduinoOTA to pass mDNS hostname to the mDNS service, just use begin function variant with hostame parameter.

About automatically copying hostname from WiFi interface, well, might be good topic to discuss... but there are other OTA implementations (for example you can have Ethernet controller and then there's a question which hostname should be used for mDNS).

**Reply by hmueller01** [reply]
Tx for the quick reply.
I think begin() is not valid, as begin is defined in ArduinoOTA:
`void begin(bool useMDNS = true);`

Have to use ArduinoOTA.setHostname("myESP")
```
//Sets the device hostname. Default esp8266-xxxxxx
    void setHostname(const char *hostname);
```
Right. I was only thinking about WiFi ... so using `WiFi.getHostname()` might not be a final solution.

But without this discussion I would never have found the source of this behavior ... (and I doubt others would).
And maybe it is good to rethink the current solution ...

**Reply by d-a-v** [reply]
DNS hostname and mDNS hostname are independant and unrelated on the network.

The first is given by the DNS server which is often a service proposed by the DHCP server on home routers.
The hostname "offered" by the esp (= dhcp client) is that one.
For this reason, `WiFi.hostname("name")` should be called before starting `WiFi.begin()` which sends the hostname proposal to the DHCP (and possibly DNS) server.

This is often difficult to manage for users (user knowledge or configuration options out of reach in some network infra).
So mDNS has been created which requires no user configuration in network servers / services.
There is no "network mDNS server". There are only devices talking each other and agreeing on name unicity on the LAN.

> But in my router I see `esp8266-6a4c81`

You router seems to show you the mDNS name configured by Arduino OTA.
If you want to change it, call `ArduinoOTA.setHostname("name")` before `ArduinoOTA.begin()`.

On the contrary my router shows me the Network name (DNS).
My computer resolve both with DNS service and mDNS service. On linux, priority is defined by `/etc/nsswitch.conf`. Don't ask me about windows.

So best practice would be:
```cpp
WiFi.hostname("name");
ArduinoOTA.setHostname("name");
WiFi.begin();
ArduinoOTA.begin();
```

> About automatically copying hostname from WiFi interface, well, might be good topic to discuss...

True.
esp8266 ArduinoOTA API has not changed for a while and it could take WiFi hostname instead of `esp8266-xxxxx` by default when uninitialized. In the past, default WiFi hostname format was not network compliant, but now it is.

**Reply by cziter15** [reply]
> Tx for the quick reply. I think begin() is not valid, as begin is defined in ArduinoOTA

My bad. Yeah, I meant setHostname before begin.

**Comment by cziter15** [comment, marked as answer]
TL;DR

- Hostname should be set **before calling WiFi.begin()**, preferably **with** WiFi **mode set to WIFI_STA (and WIFI_OFF on ESP32).**
- The **hostname** set by WiFi.setHostname **has nothing to** do with **mDNS hostname**.
- Your **code shouldn't stuck in loop(), even when you call handle() and similar methods manually** as some services are handled after returning from loop function. **Calling delay(1) or yield() is not a magic solution.** See: https://github.com/esp8266/Arduino/discussions/8816#discussioncomment-4767060
